File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/unstructured-logging.ts
#!/usr/bin/env bun
import type { Node, SourceFile } from "typescript@6.0.3";
import {
	createSourceFile,
	forEachChild,
	isCallExpression,
	isIdentifier,
	isPropertyAccessExpression,
	SyntaxKind,
} from "typescript@6.0.3";
import type {
	Finding,
	HarnessCheckRule,
	HarnessManifest,
	RuleContext,
} from "../harness-check-rule";

/**
 * Forbid unstructured logging (console.log, console.error, etc.).
 */
export const unstructuredLoggingRule = (
	ctx: RuleContext,
): HarnessCheckRule => ({
	category: "unstructuredLogging",
	applies(_manifest: HarnessManifest): boolean {
		return true;
	},

	validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
		return ctx
			.stackSources(manifest, "unstructuredLogging", "typescript")
			.map((file) => ({ file, text: ctx.read(file) }))
			.filter(({ text }) => text !== "")
			.flatMap(({ file, text }) => {
				let sourceFile: SourceFile;
				try {
					sourceFile = createSourceFile(
						file,
						text,
						SyntaxKind.LatestVersion,
						true,
					);
				} catch {
					return [
						{
							severity: ctx.severityOf(manifest, "unstructuredLogging"),
							category: "unstructuredLogging",
							message: `failed to parse TypeScript: ${file}`,
						},
					];
				}
				const findings: Finding[] = [];
				const logMethods = ["log", "error", "warn", "info", "debug"];
				const visit = (node: Node): void => {
					if (
						isCallExpression(node) &&
						isPropertyAccessExpression(node.expression) &&
						isIdentifier(node.expression.expression) &&
						node.expression.expression.text === "console"
					) {
						const methodName = node.expression.name?.text;
						if (methodName && logMethods.includes(methodName)) {
							const { line } = sourceFile.getLineAndCharacterOfPosition(
								node.getStart(sourceFile),
							);
							findings.push({
								severity: ctx.severityOf(manifest, "unstructuredLogging"),
								category: "unstructuredLogging",
								message: `${file}:${line + 1}: unstructured logging \`console.${methodName}\`; use structured logger`,
							});
						}
					}
					forEachChild(node, visit);
				};
				visit(sourceFile);
				return findings;
			});
	},
});

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/unstructured_logging.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Forbid unstructured logging rule."""

from collections.abc import Iterable

from pathlib import Path

import libcst as cst

from harness_check_rule import Finding, HarnessCheckRule

from .utils import parse_python, relative, severity_for, stack_sources

class UnstructuredLoggingRule(HarnessCheckRule):
    """Validate unstructuredLogging check."""

    category = "unstructuredLogging"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        return section.get("enabled", True) is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate unstructuredLogging check."""
        severity = severity_for(manifest, self.category)
        category = self.category
        sources = stack_sources(root, manifest, self.category)
        class PrintFinder(cst.CSTVisitor):
            METADATA_DEPENDENCIES = (cst.metadata.PositionProvider,)
            def __init__(self, rel_path: str) -> None:
                super().__init__()
                self.findings: list[Finding] = []
                self.rel_path = rel_path
            def visit_Call(self, node: cst.Call) -> bool:
                if isinstance(node.func, cst.Name):
                    if node.func.value == "print":
                        pos = self.get_metadata(cst.metadata.PositionProvider, node)
                        self.findings.append(Finding(
                            severity,
                            category,
                            f"{self.rel_path}:{pos.start.line}: unstructured logging `print`; use structured logger",
                        ))
                return True
        result = []
        for path in sources:
            tree, error = parse_python(path)
            if error is not None:
                result.append(Finding(
                    severity,
                    category,
                    f"{relative(path)}: syntax error: {error}",
                ))
                continue
            wrapper = cst.MetadataWrapper(tree)
            visitor = PrintFinder(relative(path))
            wrapper.visit(visitor)
            result.extend(visitor.findings)
        return result


RULE: HarnessCheckRule = UnstructuredLoggingRule()
